core/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java

```diff
@@ -222,6 +223,59 @@ public GroupShardsIterator allAssignedShardsGrouped(String[] indices, boolean in
         return new GroupShardsIterator(set);
     }
 
+    private final static Predicate<ShardRouting> ACTIVE_PREDICATE = new Predicate<ShardRouting>() {
+        @Override
+        public boolean apply(ShardRouting shardRouting) {
+            return shardRouting.active();
+        }
+    };
+
+    private final static Predicate<ShardRouting> ASSIGNED_PREDICATE = new Predicate<ShardRouting>() {
+        @Override
+        public boolean apply(ShardRouting shardRouting) {
+            return shardRouting.assignedToNode();
+        }
+    };
+
+    public ShardsIterator allAssignedShards(String[] indices, boolean includeEmpty) {
+        return allAssignedShards(indices, includeEmpty, false);
+    }
+
+    public ShardsIterator allAssignedShards(String[] indices, boolean includeEmpty, boolean includeRelocationTargets) {
+        return allShardsSatisfyingPredicate(indices, includeEmpty, includeRelocationTargets, ASSIGNED_PREDICATE);
+    }
+
+    public ShardsIterator allActiveShards(String[] indices, boolean includeEmpty) {
+        return allActiveShards(indices, includeEmpty, false);
+    }
+
+    public ShardsIterator allActiveShards(String[] indices, boolean includeEmpty, boolean includeRelocationTargets) {
+        return allShardsSatisfyingPredicate(indices, includeEmpty, includeRelocationTargets, ACTIVE_PREDICATE);
+    }
+
+    private ShardsIterator allShardsSatisfyingPredicate(String[] indices, boolean includeEmpty, boolean includeReloactionTargets, Predicate<ShardRouting> predicate) {
+        // use list here since we need to maintain identity across shards
+        List<ShardRouting> shards = new ArrayList<>();
+        for (String index : indices) {
+            IndexRoutingTable indexRoutingTable = index(index);
+            if (indexRoutingTable == null) {
+                continue;
+                // we simply ignore indices that don't exists (make sense for operations that use it currently)
+            }
+            for (IndexShardRoutingTable indexShardRoutingTable : indexRoutingTable) {
+                for (ShardRouting shardRouting : indexShardRoutingTable) {
+                    if (predicate.apply(shardRouting) || includeEmpty) {
```

Comment:
Done this way, the predicate method for a plain `ShardsIterator` isn't needed at this time. If you think I should add it anyway, I can but for now I've left it out. Otherwise, I've addressed this comment.